File: app/schemas/ap_metrics.py
"""Response shape for the AP dashboard."""

from datetime import datetime

from pydantic import BaseModel


class ExceptionCount(BaseModel):
    code: str
    count: int


class AgingBucket(BaseModel):
    bucket: str
    count: int


class RecentRun(BaseModel):
    run_id: str
    invoice_ref: str | None = None
    status: str
    started_at: datetime | None = None
    duration_ms: int | None = None
    policy_version_label: str | None = None


class APMetricsResponse(BaseModel):
    invoices_processed: int
    touchless_rate: float
    pay_ready: int
    verdict_breakdown: dict[str, int]

    # Grouped by currency — never summed across them.
    money_protected: dict[str, float]
    spend_under_review: dict[str, float]

    exceptions_by_type: list[ExceptionCount]
    workbench_open: int
    workbench_resolved: int
    workbench_priority: dict[str, int]
    exception_aging: list[AgingBucket]

    average_confidence: float | None = None
    total_runs: int
    average_run_ms: int | None = None
    recent_runs: list[RecentRun]
    last_decision_at: datetime | None = None

File: frontend/src/types/ap-metrics.ts
export interface APExceptionCount {
  code: string
  count: number
}

export interface APAgingBucket {
  bucket: string
  count: number
}

export interface APRecentRun {
  run_id: string
  invoice_ref: string | null
  status: string
  started_at: string | null
  duration_ms: number | null
  policy_version_label: string | null
}

export interface APMetrics {
  invoices_processed: number
  touchless_rate: number
  pay_ready: number
  verdict_breakdown: Record<string, number>

  /** Grouped by currency. Never summed across currencies. */
  money_protected: Record<string, number>
  spend_under_review: Record<string, number>

  exceptions_by_type: APExceptionCount[]
  workbench_open: number
  workbench_resolved: number
  workbench_priority: Record<string, number>
  exception_aging: APAgingBucket[]

  average_confidence: number | null
  total_runs: number
  average_run_ms: number | null
  recent_runs: APRecentRun[]
  last_decision_at: string | null
}
